Source/FieldSolver/ElectrostaticSolver.cpp

```diff
@@ -358,6 +368,122 @@ WarpX::computePhiCartesian (const amrex::Vector<std::unique_ptr<amrex::MultiFab>
 }
 #endif
 
+/* \bried Set Dirichlet boundary conditions for the electrostatic solver.
+
+    The given potential's values are fixed on the boundaries of the given
+    dimension according to the desired values from the simulation input file,
+    geometry.potential_lo and geometry.potential_hi.
+
+   \param[inout] phi The electrostatic potential
+   \param[in] idim The dimension for which the Dirichlet boundary condition is set
+*/
+void
+WarpX::setDirichletBC(amrex::Vector<std::unique_ptr<amrex::MultiFab> >& phi,
+                      const int idim) const
+{
+    Real potential_lo;
+    Real potential_hi;
+
+    getBoundaryPotentials(idim, potential_lo, potential_hi);
+
+    // Print() << "Potential lo = " << potential_lo << "\n";
+    // Print() << "Potential hi = " << potential_hi << "\n";
+
+    // loop over all multigrid levels and set the boundary values
+    for (int lev=0; lev <= max_level; lev++) {
+        for (MFIter mfi(*phi[lev], TilingIfNotGPU()); mfi.isValid(); ++mfi) {
+            FArrayBox& fab = (*phi[0])[mfi]; //
+            Array4<Real> const& phi_gh = fab.array(); // phi include ghost cells
+            const auto lo = lbound(phi_gh);
+            const auto hi = ubound(phi_gh);
+
+            //TODO use AMREX_GPU_DEVICE for acceleration
+            // The boundary values are set at points lo.idim + 1 and hi.idim -1
+            // since phi_gh includes ghost cells
+            if (idim == 0){
+                for (int k=lo.z; k<=hi.z; k++) {
+                for (int j=lo.y; j<=hi.y; j++) {
+                    phi_gh(lo.x+1,j,k) = potential_lo;
+                    phi_gh(hi.x-1,j,k) = potential_hi;
+                }}
+            }
+            if (idim == 1){
+                for (int k=lo.z; k<=hi.z; k++) {
+                for (int i=lo.x; i<=hi.x; i++) {
+                    phi_gh(i,lo.y+1,k) = potential_lo;
+                    phi_gh(i,hi.y-1,k) = potential_hi;
+                }}
+            }
+            if (idim == 2){
+                for (int j=lo.y; j<=hi.y; j++) {
+                for (int i=lo.x; i<=hi.x; i++) {
+                    phi_gh(i,j,lo.z+1) = potential_lo;
+                    phi_gh(i,j,hi.z-1) = potential_hi;
+                }}
+            }
+    }} // lev & MFIter
+
+    /*
+    Wei-Ting's implementation with GPU acceleration:
```

Comment:
Not that this is live code, but it looks like it assigns values for the *whole* domain, and not just the boundary cells. Besides being locally inefficient, it could hypothetically mess with the iterative solver, since it potentially replaces the most recent solution with the straight linear interpolation, making it start to converge from scratch.